mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp

```diff
@@ -367,10 +378,52 @@ static LogicalResult defineDeclaredSymbols(Block &block, ModuleOp definitions) {
       }
     }
 
-    OpBuilder builder(&op);
-    builder.setInsertionPoint(&op);
+    OpBuilder builder(symbol);
+    builder.setInsertionPoint(symbol);
     builder.clone(*externalSymbol);
     symbol->erase();
+
+    LLVM_DEBUG(DBGS() << "scanning definition of @"
+                      << externalSymbolFunc.getNameAttr().getValue()
+                      << " for symbol usages\n");
+    externalSymbolFunc.walk([&](CallOpInterface callOp) {
+      LLVM_DEBUG(DBGS() << "  found symbol usage in:\n" << callOp << "\n");
+      CallInterfaceCallable callable = callOp.getCallableForCallee();
+      if (!isa<SymbolRefAttr>(callable)) {
+        LLVM_DEBUG(DBGS() << "    not a 'SymbolRefAttr'\n");
+        return WalkResult::advance();
+      }
+
+      StringRef callableSymbol =
+          cast<SymbolRefAttr>(callable).getLeafReference();
+      LLVM_DEBUG(DBGS() << "    looking for @" << callableSymbol
+                        << " in definitions: ");
+
+      Operation *callableOp = definitionsSymbolTable.lookup(callableSymbol);
+      if (!isa<SymbolRefAttr>(callable)) {
+        LLVM_DEBUG(llvm::dbgs() << "not found\n");
+        return WalkResult::advance();
+      }
+      LLVM_DEBUG(llvm::dbgs() << "found " << callableOp << " from "
+                              << callableOp->getLoc() << "\n");
+
+      if (!block.getParent() || !block.getParent()->getParentOp()) {
+        LLVM_DEBUG(DBGS() << "could not get parent of provided block");
+        return WalkResult::advance();
+      }
+
+      SymbolTable targetSymbolTable(block.getParent()->getParentOp());
+      if (targetSymbolTable.lookup(callableSymbol)) {
+        LLVM_DEBUG(DBGS() << "    symbol @" << callableSymbol
+                          << " already present in target\n");
+        return WalkResult::advance();
+      }
+
```

Comment:
I am not perfectly happy with this part yet. I am wondering whether `defineDeclaredSymbols` shouldn't accept the parent op right away rather than a block. If that isn't acceptable for some reason, I wonder whether this shouldn't fail rather just outputting a silent debug message.

Reply:
We need to be careful here, as this basically a proto-linker. One low-tech solution could be to just clone all symbols from the library file(s) into the main one, not just the declared/referenced ones. The difficult issue in any case is name clashes between files. AFAIR, `named_sequence` doesn't have the visibility attribute and it probably should for this purpose. It would explicitly tell us where renaming a symbol to avoid the name clash is acceptable (private symbol) and when it isn't. When it isn't, the process needs to run in three stages: (1) collect all definitions (i.e., with bodies) of named sequences that need to be cloned without actually cloning them + check that there is exactly one definition of each if renaming is not allowed; (2) if needed and allowed, rename private symbols and update their uses in each module individually; (3) actually clone (also let's consider moving them instead).

Limiting the process to only the symbols that are actually needed is an optimization. We can do it, but can also punt. In the general case, we need something like connected components for all named sequences in the main file as it may also have unused declarations. There's `CallGraph` for that and LLVM has algorithms to compute connected components so no need to write this manually.

Reply:
Yeah, this *is* kind of a linker. I am actually wondering if it shouldn't live in a more central place?

But, boy, it's the third time that I realize that this is more involved than I originally thought :P But I think I understand the issue and should be able to fix it.

Reply:
Eventually, it could become a more generic utility. But I wouldn't go out of my way right now as I'm not aware of any other use case, just keeping it relatively separable by being interface-based (symbol tables, call graph) should be sufficient.

Reply:
`NamedSequenceOp` [has the](https://github.com/llvm/llvm-project/blob/2f8690b1e2b/mlir/include/mlir/Dialect/Transform/IR/TransformOps.td#L806) `FunctionOpInterface`, which [inherits from](https://github.com/llvm/llvm-project/blob/2f8690b1e2b/mlir/include/mlir/Interfaces/FunctionInterfaces.td#L25) `SymbolOpInterface`, which is where `getVisibility` [is declared/defined](https://github.com/llvm/llvm-project/blob/2f8690b1e2b/mlir/include/mlir/IR/SymbolInterfaces.td#L49-L54), so named sequences already have the visibility attribute as it seems.

Reply:
OK, I think that you propose makes sense. Linking in all symbols instead of just the referenced once is something that I need for #67120.

One question, though: `SymbolOpInterface` alone does not know the concept of "external" (aka the distinction between definition vs declaration); only `FunctionOpInterface` does. However, the previous implementation of `defineDeclaredSymbols` [tested](https://github.com/llvm/llvm-project/blob/2f8690b1e2b/mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp#L316-L320) for `SymbolOpInterface` and `symbol->getNumRegions() == 1 && !symbol->getRegion(0).empty()`, which corresponds to the [definition of `isExternal`](https://github.com/llvm/llvm-project/blob/2f8690b1e2b/mlir/include/mlir/Interfaces/FunctionInterfaces.td#L163) in `FunctionOpInterface`. I think the more general case to deal with that is to implement the case that allows external functions (aka declarations) for the special case `FunctionOpInterface` and to not allow clashes of public symbols otherwise. Do you agree?

Reply:
Sounds good as a first approximation. I suspect that one will have to lift the declaration/definition distinction out of `FunctionOpInterface` into either `SymbolOpInterface` or a new interface as it makes sense for other kinds of symbols such as memref globals. We don't care about these for our use case, so let's stick with functions.

Reply:
> Limiting the process to only the symbols that are actually needed is an optimization. We can do it, but can also punt. In the general case, we need something like connected components for all named sequences in the main file as it may also have unused declarations. There's `CallGraph` for that and LLVM has algorithms to compute connected components so no need to write this manually.

I briefly started to think about this. The call graph analysis works on `CallOpInterface` and `CallableOpInterface`, which I think should be enough for `transform.include` and `transform.sequence` at the moment. However, if some op used a symbol defined in the library module that does *not* implement the `CallableOpInterface`, we wouldn't capture that relationship with the call graph analysis. I don't know of any (transform) op that does that currently, so it may not be an issue. For full generality, wouldn't we have to build a "used by" graph of symbols (or even a mixture of that and the call graph)?

Reply:
In the general case, yes, we would need to care about things like `addressof` that are not a call. I don't have an immediate plan for having something like that in the transform dialect.